File: agents/python/shared/guardrails/moderation.py
"""对模型输出执行粗粒度的本地内容审核。

与净化输入中的恶意指令不同，这里检查模型自己生成的自伤、暴力、
仇恨、骚扰和色情等内容。即使抵御了注入，自由生成文本仍可能有风险。

全部是纯函数，不执行 I/O、模型或外部 API 调用。少量规则优先降低
误报，不追求完整召回；这不是训练分类器，间接措辞可能漏检。
详见 docs/concepts/10-guardrails.md 的分层防护说明。
"""

from __future__ import annotations

import re
from enum import StrEnum


class ModerationCategory(StrEnum):
    SELF_HARM = "self_harm"
    VIOLENCE = "violence"
    HATE_HARASSMENT = "hate_harassment"
    SEXUAL = "sexual"


_PATTERNS: dict[ModerationCategory, tuple[re.Pattern[str], ...]] = {
    ModerationCategory.SELF_HARM: (
        re.compile(r"\bkill\s+(?:myself|yourself)\b", re.I),
        re.compile(r"\b(?:commit|committing)\s+suicide\b", re.I),
        re.compile(r"\bways?\s+to\s+(?:end|take)\s+(?:my|your|his|her|their)\s+(?:own\s+)?life\b", re.I),
        re.compile(r"\bself[\s-]harm\b", re.I),
    ),
    ModerationCategory.VIOLENCE: (
        re.compile(r"\bhow\s+to\s+(?:build|make)\s+a\s+(?:bomb|explosive|weapon)\b", re.I),
        re.compile(r"\bi\s+(?:will|'ll|am\s+going\s+to)\s+kill\s+you\b", re.I),
        re.compile(r"\bmass\s+shooting\b", re.I),
    ),
    ModerationCategory.HATE_HARASSMENT: (
        re.compile(r"\ball\s+\w+\s+(?:people\s+)?(?:are|should\s+be)\s+(?:killed|exterminated|eliminated)\b", re.I),
        re.compile(r"\byou'?re\s+(?:a\s+)?(?:worthless|subhuman)\b", re.I),
    ),
    ModerationCategory.SEXUAL: (
        re.compile(r"\bsexually\s+explicit\s+(?:content|description|story)\b", re.I),
        re.compile(r"\bchild\s+sexual\s+abuse\b", re.I),
    ),
}


def classify(text: str) -> set[ModerationCategory]:
    """返回文本命中的所有类别；空集合表示未命中规则。"""
    hits: set[ModerationCategory] = set()
    for category, patterns in _PATTERNS.items():
        if any(p.search(text) for p in patterns):
            hits.add(category)
    return hits

File: agents/dotnet/src/ECommerceAgents.Shared/Guardrails/Moderation.cs
using System.Text.RegularExpressions;

namespace ECommerceAgents.Shared.Guardrails;

/// <summary>Content-policy categories a piece of text can be flagged under.</summary>
public enum ModerationCategory
{
    SelfHarm,
    Violence,
    HateHarassment,
    Sexual,
}

/// <summary>
/// Coarse local content-moderation classifier for outbound agent text — the
/// .NET twin of Python's <c>shared/guardrails/moderation.py</c>.
/// </summary>
/// <remarks>
/// Distinct from <see cref="Sanitize"/>: sanitization defangs adversarial
/// <em>instructions</em> hiding inside untrusted input (a review body
/// telling the model to "ignore previous instructions") so they never
/// influence the next turn. This class classifies the model's own
/// <em>output</em> text against content-policy categories (self-harm,
/// violence, hate/harassment, sexual content) — a different problem. An
/// agent that correctly resists every injection attempt can still generate
/// harmful text on its own.
///
/// Pure functions: no I/O, no LLM call, no external API. Deliberately a
/// small set of high-precision phrase patterns, mirroring
/// <see cref="Sanitize"/>'s own philosophy (low false-positive over
/// exhaustive recall) — this is a coarse first-pass filter, not a trained
/// classifier. Patterns copied verbatim from the Python source so the two
/// stacks flag the same text.
/// </remarks>
public static class Moderation
{
    private static readonly Dictionary<ModerationCategory, Regex[]> Patterns = new()
    {
        [ModerationCategory.SelfHarm] =
        [
            new(@"\bkill\s+(?:myself|yourself)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\b(?:commit|committing)\s+suicide\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\bways?\s+to\s+(?:end|take)\s+(?:my|your|his|her|their)\s+(?:own\s+)?life\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\bself[\s-]harm\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        ],
        [ModerationCategory.Violence] =
        [
            new(@"\bhow\s+to\s+(?:build|make)\s+a\s+(?:bomb|explosive|weapon)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\bi\s+(?:will|'ll|am\s+going\s+to)\s+kill\s+you\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\bmass\s+shooting\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        ],
        [ModerationCategory.HateHarassment] =
        [
            new(@"\ball\s+\w+\s+(?:people\s+)?(?:are|should\s+be)\s+(?:killed|exterminated|eliminated)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\byou'?re\s+(?:a\s+)?(?:worthless|subhuman)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        ],
        [ModerationCategory.Sexual] =
        [
            new(@"\bsexually\s+explicit\s+(?:content|description|story)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new(@"\bchild\s+sexual\s+abuse\b", RegexOptions.IgnoreCase | RegexOptions.Compiled),
        ],
    };

    /// <summary>Returns every category whose patterns match <paramref name="text"/>. Empty = clean.</summary>
    public static HashSet<ModerationCategory> Classify(string? text)
    {
        var hits = new HashSet<ModerationCategory>();
        if (string.IsNullOrEmpty(text))
        {
            return hits;
        }

        foreach (var (category, patterns) in Patterns)
        {
            if (patterns.Any(p => p.IsMatch(text)))
            {
                hits.Add(category);
            }
        }
        return hits;
    }
}
